### PR TITLE
Multiview updates

### PR DESCRIPTION
**Pull request by admshao**
- Clean up
- Small performance improvement
- Adjusted label size
- Readjust multiview on video settings change
- New multiview settings
- New multiview layout

screenshot from 2018-04-14 18-45-33
screenshot from 2018-04-14 18-45-04
screenshot from 2018-04-14 18-44-00
screenshot from 2018-04-14 19-08-46

### CODE REVIEW

File: UI/data/locale/en-US.ini [hunk 1]
```diff
@@ -570,11 +570,16 @@ Basic.Settings.General.SystemTrayHideMinimize="Always minimize to system tray in
 Basic.Settings.General.SaveProjectors="Save projectors on exit"
 Basic.Settings.General.SwitchOnDoubleClick="Transition to scene when double-clicked"
 Basic.Settings.General.StudioPortraitLayout="Enable portrait/vertical layout"
+Basic.Settings.General.Multiview="Multiview"
+Basic.Settings.General.Multiview.MouseSwitch="Change scenes with clicks"
```
Comment (WizardCM) [inline UI/data/locale/en-US.ini]: I'd probably name this "Click to switch between scenes"

File: UI/data/locale/en-US.ini [hunk 2]
```diff
@@ -570,11 +570,16 @@ Basic.Settings.General.SystemTrayHideMinimize="Always minimize to system tray in
...
+Basic.Settings.General.Multiview.DrawSourceNames="Draw sources name"
```
Comment (WizardCM) [inline UI/data/locale/en-US.ini]: Based on my understanding, this is related to scenes rather than sources, yeah?

"Show scene names" might suit better.

File: UI/data/locale/en-US.ini [hunk 3]
```diff
@@ -570,11 +570,16 @@ Basic.Settings.General.SystemTrayHideMinimize="Always minimize to system tray in
...
+Basic.Settings.General.Multiview.DrawSafeAreas="Draw safe areas (EBU R 95)"
 Basic.Settings.General.MultiviewLayout="Multiview Layout"
-Basic.Settings.General.MultiviewLayout.Horizontal.Top="Horizontal, Top"
-Basic.Settings.General.MultiviewLayout.Horizontal.Bottom="Horizontal, Bottom"
-Basic.Settings.General.MultiviewLayout.Vertical.Left="Vertical, Left"
-Basic.Settings.General.MultiviewLayout.Vertical.Right="Vertical, Right"
+Basic.Settings.General.MultiviewLayout.Horizontal.Top="Horizontal, Top (2X4X4)"
```
Comment (WizardCM) [inline UI/data/locale/en-US.ini]: This might come down to personal preference, but would small 'x' improve readability? 2x4x4 vs 2X4X4

File: UI/data/locale/en-US.ini [hunk 4]
```diff
@@ -570,11 +570,16 @@ Basic.Settings.General.SystemTrayHideMinimize="Always minimize to system tray in
...
+Basic.Settings.General.MultiviewLayout.Horizontal.Bottom="Horizontal, Bottom (4X4X2)"
+Basic.Settings.General.MultiviewLayout.Vertical.Left="Vertical, Left (2X4X4)"
+Basic.Settings.General.MultiviewLayout.Vertical.Right="Vertical, Right (4X4X2)"
+Basic.Settings.General.MultiviewLayout.Horizontal.Extended.Top="Horizontal, Top (1_2_1X6X4)"
```
Comment (WizardCM) [inline UI/data/locale/en-US.ini]: The layout measurements here would be confusing to most users, I'm not exactly sure what it means just by reading it. Not sure how it could be improved.

File: UI/data/locale/en-US.ini (hunk 2 above)
Comment (admshao) [reply UI/data/locale/en-US.ini]: That is correct. Will update.

File: UI/data/locale/en-US.ini (hunk 4 above)
Comment (admshao) [reply UI/data/locale/en-US.ini]: Would change the "(aXbXc)" to juse "8 scenes"/"24 scenes" work better?

File: UI/data/locale/en-US.ini (hunk 1 above)
Comment (admshao) [reply UI/data/locale/en-US.ini]: Will update.

File: UI/data/locale/en-US.ini (hunk 4 above)
Comment (WizardCM) [reply UI/data/locale/en-US.ini]: I think it would, as most people won't care about the size/shape of the grid but how many scenes they get out of it. That's a good idea.

File: UI/window-basic-settings.cpp [hunk 5]
```diff
@@ -1102,28 +1102,20 @@ void OBSBasicSettings::LoadGeneralSettings()
...
-		ui->multiviewLayout->setCurrentIndex(1);
-	else if (astrcmpi(multiviewLayoutText, "verticalleft") == 0)
-		ui->multiviewLayout->setCurrentIndex(2);
-	else if (astrcmpi(multiviewLayoutText, "verticalright") == 0)
-		ui->multiviewLayout->setCurrentIndex(3);
-	else
-		ui->multiviewLayout->setCurrentIndex(0);
+	ui->multiviewLayout->setCurrentIndex(
+			config_get_int(GetGlobalConfig(), "BasicWindow",
+					"MultiviewLayout"));
```
Comment (jp9000) [inline UI/window-basic-settings.cpp]: You're changing what a config file parameter expects here.  Previously, the [BasicWindow] MultiviewLayout variable was expected to be a string, and now you're sort of abruptly changing it to an int, basically breaking a user's existing config.

First it has to be noted that the reason why list index values aren't usually used as config values is because if you add/change something in that list box, it breaks configuration, hence why strings are typically used to represent their values.  Basically, it was done to prevent the chance of breaking configuration.

While an int is technically easier to use here, it's not necessarily safe due to that reason.

Changing it from expecting a string value to expecting an int value just breaks previous config values for this int, which is not typically a good practice to do.

1.) You can do this and allow it to break configuration
2.) You can use the string values instead.  There are probably easier ways to map string -> int values than this.  If possible, I would recommend to try to make it continue to use a string if it works with the current configuration.

File: UI/window-projector.hpp [hunk 6]
```diff
@@ -49,6 +49,13 @@ class OBSProjector : public OBSQTDisplay {
 	gs_vertbuffer_t *rightLine = nullptr;
 	bool ready = false;
 
+	// argb colors
+	static const uint32_t outerColor     = 0xFFD0D0D0;
```
Comment (jp9000) [inline UI/window-projector.hpp]: Indent here doesn't appear to be the same as the ones below it.  Just a totally minor and insignificant nitpick.

File: UI/window-projector.cpp [hunk 7]
```diff
@@ -438,12 +429,14 @@ void OBSProjector::OBSRenderMultiview(void *data, uint32_t cx, uint32_t cy)
 		}
 	};
 
-	/* ----------------------------- */
-	/* draw sources                  */
-
```
Comment (jp9000) [inline UI/window-projector.cpp]: I'd probably not recommend removing these sort of code segmentation comments.  These comments were put in here to make a larger function like this a bit more readable.

File: UI/window-projector.cpp [hunk 8]
```diff
@@ -451,17 +451,17 @@ void OBSProjector::OBSRenderMultiview(void *data, uint32_t cx, uint32_t cy)
 		qiX = sourceX + 4.0f;
 		qiY = sourceY + 4.0f;
 
-		/* ----------- */
```
Comment (jp9000) [inline UI/window-projector.cpp]: Again, just wanted to note that you removed the segmentation comment here.  It also doesn't appear to be related to this particular commit.

File: UI/window-projector.hpp [hunk 9]
```diff
@@ -40,16 +40,21 @@ class OBSProjector : public OBSQTDisplay {
...
+	float lineLength                = 0.1f;
+	// Rec. ITU-R BT.1848-1 / EBU R 95
+	float actionSafePercentage      = 0.035f; // 3.5%
+	float graphicsSafePercentage    = 0.05f; // 5.0%
+	float fourByThreeSafePercentage = 0.1625f; // 16.25%
 	bool ready = false;
 
 	// argb colors
-	static const uint32_t outerColor     = 0xFFD0D0D0;
+	static const uint32_t outerColor      = 0xFFD0D0D0;
```
Comment (jp9000) [inline UI/window-projector.hpp]: You used this commit to correct a previous unrelated commit here.  Something that you want to avoid for ideal review diffs.

File: UI/window-projector.cpp [hunk 10]
```diff
@@ -831,7 +828,6 @@ void OBSProjector::UpdateMultiview()
 
 	struct obs_video_info ovi;
 	obs_get_video_info(&ovi);
-
```
Comment (jp9000) [inline UI/window-projector.cpp]: A strange little removal of a newline here.

File: UI/window-projector.cpp [hunk 11]
```diff
@@ -375,86 +354,87 @@ void OBSProjector::OBSRenderMultiview(void *data, uint32_t cx, uint32_t cy)
...
+				window->sourceX += window->halfCX;
+				window->labelX += window->halfCX;
 			}
 		}
 	};
 
-	auto paintAreaWithColor = [&](float tx, float ty, float cx, float cy,
-			uint32_t color)
+	auto paintAreaWithColor = [&](float tx, float ty, float cx,
+			float cy, uint32_t color)
```
Comment (jp9000) [inline UI/window-projector.cpp]: Noticed here that you changed the parameter placement here unnecessarily.  The "float cy" parameter wasn't particularly going past the column limit.

File: UI/window-projector.cpp [hunk 12]
```diff
@@ -333,33 +333,33 @@ void OBSProjector::OBSRenderMultiview(void *data, uint32_t cx, uint32_t cy)
...
 			sourceX = 0;
 			sourceY = (i / 2 ) * quarterCY;
 			if (i % 2 != 0)
 				sourceX = quarterCX;
 			break;
-		case MultiviewLayout::HORIZONTAL_BOTTOM:
+		case MultiviewLayout::HORIZONTAL_BOTTOM_8_SCENES:
 			if (i < 4) {
-				sourceX = (float(i) * quarterCX);
+				sourceX = i * quarterCX;
```
Comment (jp9000) [inline UI/window-projector.cpp]: I noticed you unnecessarily changing the explicit casts all over here.  I did explicit casts here to make the reader of the code more aware that the variable is not a float, and to prevent any conversion warnings (if any).  I admit it's become a habit due to microsoft's compiler design, but still, I feel these changes were an unnecessary/unrealted part of this commit/diff.

File: UI/window-projector.cpp [hunk 13]
```diff
@@ -659,22 +660,22 @@ static int getSourceByPosition(int x, int y)
 	int     minY  = 0;
 	int     maxX  = cx;
 	int     maxY  = cy;
-	int     halfX = cx / 2;
-	int     halfY = cy / 2;
+	int     pvwpgmX = cx / 2;
+	int     pvwpgmY = cy / 2;
```
Comment (jp9000) [inline UI/window-projector.cpp]: I am unsure why these names were changed for these variables.  The name you named them to makes them seemingly less readable, and add unnecessary changes to this commit/diff that sort of confuse me here, and have no mention in the commit message.

File: UI/window-projector.cpp [hunk 14]
```diff
@@ -350,82 +353,98 @@ void OBSProjector::OBSRenderMultiview(void *data, uint32_t cx, uint32_t cy)
...
 		switch (multiviewLayout) {
+		case MultiviewLayout::HORIZONTAL_TOP_24_SCENES:
+			window->sourceX = (i % 6) * window->scenesCX;
+			window->sourceY = window->pvwprgCY +
+					(i / 6) * window->scenesCY;
+			break;
 		case MultiviewLayout::VERTICAL_LEFT_8_SCENES:
-			window->sourceX = window->halfCX;
-			window->sourceY = (i / 2 ) * window->quarterCY;
+			window->sourceX = window->pvwprgCX;
```
Comment (jp9000) [inline UI/window-projector.cpp]: Unfortunately, due to the name change of these variables, this commit also appears to be affected with sort of "unrelated" changes, or so it feels.  Perhaps I will need to consult with you on why the variable name changes occurred, but they don't appear to have been mentioned in any commit messages.

**Comment by jp9000** [comment]
Despite my review, I'll probably allow most of these things I commented on, so you don't necessarily have to change everything I commented on -- that would probably be a pain.  I'm going to give it a few more days to let you decide whether you want to act on those commit changes or not, but they're not a huge deal as long as you learn from them in the future I feel.

You're getting really good at bigger pull changes, which is not easy to do.  Your diffs are getting much better.

Ideally, what you want to do before submitting your PR is use github's compare feature for your branch, and examine each commit individually to see if they feel right to you and don't have anything out of place.  Imagine yourself in the reviewer's shoes looking at each commit to see how you would feel about the changes.

The one thing I do want you to consider changing (or solving) is the configuration value change.  You changed an ini user configuration variable from a string to an int, which breaks the user's previous configuration for that one particular setting (causing it to "reset" to a different value).  I'm unsure if this was intentional.  One way you could approach solving this is to write something that converts old values to new values, I've done that a few times.  That would allow you to just simply add one extra commit to this branch, rather than necessarily having to go back and edit your previous commit.

It's just not really a nice thing for user experience when a program breaks existing user config settings with a new version, so that's really the sole important issue of this pull request that I can see so far.

**Comment by xmetaldavex** [comment]
I'd like to use this feature. Will it be included in an update, or do I need to manually modify the files?
